### PR TITLE
Fix Concurrent Snapshot Ending And Stabilize Snapshot Finalization

### LINKED ISSUE #38226: testAbortedSnapshotDuringInitDoesNotStart fails with ClassCastException
This test is failing with ClassCastException.

```
  1> org.elasticsearch.transport.RemoteTransportException: [node_sd3][127.0.0.1:34425][internal:cluster/snapshot/update_snapshot_status]
  1> Caused by: org.elasticsearch.transport.ResponseHandlerFailureTransportException: java.lang.ClassCastException: org.elasticsearch.snapshots.SnapshotShardsService$UpdateIndexShardSnapshotStatusResponse cannot be cast to org.elasticsearch.transport.TransportResponse$Empty
  1> Caused by: java.lang.ClassCastException: org.elasticsearch.snapshots.SnapshotShardsService$UpdateIndexShardSnapshotStatusResponse cannot be cast to org.elasticsearch.transport.TransportResponse$Empty
  1> 	at org.elasticsearch.snapshots.SnapshotShardsServic
```

I can't reproduce this locally but this test failed 4 times today.

```
 ./gradlew :server:integTest \
  -Dtests.seed=F8B30D4F31180D3F \
  -Dtests.class=org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT \
  -Dtests.method="testAbortedSnapshotDuringInitDoesNotStart" \
  -Dtests.security.manager=true \
  -Dtests.locale=cs \
  -Dtests.timezone=AET \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

CI:
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1768/console
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1763/console
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=amazon/217/console
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+internalClusterTest/262/console
- https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+default-distro-bwc-tests/326/consoleText

### PR DESCRIPTION
**Pull request by original-brownbear**
* Partly extracted and inspired by https://github.com/elastic/elasticsearch/compare/master...ywelsch:snapshot-refactored?expand=1#diff-a0853be4492c052f24917b5c1464003dR975
* The problem in #38226 is that in some corner cases multiple calls to `endSnapshot` were made concurrently, leading to non-deterministic behavior (`beginSnapshot` was triggering a repository finalization while one that was triggered by a `deleteSnapshot` was already in progress)
   * Fix by:
      * Making all `endSnapshot` calls originate from the cluster state being in a "completed" state (apart from on short-circuit on initializing an empty snapshot). This forced putting the failure string into `SnapshotsInProgress.Entry`.
      * Adding deduplication logic to `endSnapshot`
* Also:
  * Streamlined the init behavior to work the same way (keep state on the `SnapshotsService` to decide which snapshot entries are stale)
* closes #38226 

-----------------------------

Note: I ran a few thousand iterations of the `SnapshotResiliencyTests` for these changes and they came back green,

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java [hunk 1]
```diff
@@ -3684,14 +3683,9 @@ public void testAbortedSnapshotDuringInitDoesNotStart() throws Exception {
 
             // The deletion must set the snapshot in the ABORTED state
             assertBusy(() -> {
-                try {
-                    SnapshotsStatusResponse status =
-                        client.admin().cluster().prepareSnapshotStatus("repository").setSnapshots("snap").get();
-                    assertThat(status.getSnapshots().iterator().next().getState(), equalTo(State.ABORTED));
-                } catch (Exception e) {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: This isn't necessary anymore, we'll never create a broken repository with this fix.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2

**Comment by original-brownbear** [comment]
@ywelsch I tried to fix this in a shorter manner (i.e. without having to make wire protocol changes to `SnapshotsInProgress`), but I eventually decided against it:
* The resiliency tests could not catch the problem here because it came from 2 concurrent threads in the snapshot worker pool interfering with each other (and obviously we only execute those sequentially in the deterministic task queue)
* Simply adding the deduplicating of `endSnapshot` calls introduces some new concurrency spots to think about and I figured it's safer to go with this approach (in general the one you took in your branch as well) and at least completely eliminate the risk of stuck snapshots where all shards have finished and get away from the brittle `if (newMaster)` style logic.

Take a look when you have a chance (diff isn't so large with whitespaces ignored :)).

**Comment by original-brownbear** [comment]
test failure is due to #38412

**Review by ywelsch — APPROVED** [review]

**Comment by original-brownbear** [comment]
@ywelsch thanks!

**Pull request merged**